### ISSUE TITLE
lava.lib.dl.netx.hdf5 imports Convolutional Layers incorrectly

### ISSUE THREAD

**Issue by Karl-Vetter** [lava-nc/lava-dl#260]
**Description**
When loading a network trained with slayer using the hdf5 importing functionality the kernels get transposed leading to incorrect outputs.

**Steps to reproduce the behavior:**
Use this code to train a network on a dummy-dataset using slayer and save it using h5py
```python
import numpy as np
import lava.lib.dl.slayer as slayer
import torch
import lava
import h5py



neuronparams_cuba = {}
neuronparams_cuba["threshold"] = 0.5

neuronparams_cuba["current_decay"] = 1
neuronparams_cuba["voltage_decay"] = 0
neuronparams_cuba["shared_param"] = False
neuronparams_cuba["scale"] = 64
neuronparams_cuba["requires_grad"] = True
neuronparams_cuba["scale_grad"] = 0.1



neuronparams_cuba_inp = {}
neuronparams_cuba_inp["threshold"] = 0.1
neuronparams_cuba_inp["scale"] = 64

neuronparams_cuba_inp["current_decay"] = 1
neuronparams_cuba_inp["voltage_decay"] = 0


def customClassifier(output):
    return torch.argmax(output, dim = 1)



# Defining simple dummy dataset to demonstrate the problem

basic_input = []
basic_input.append(np.array([[1,0,0], [0,0,0], [0,0,0]]))
basic_input.append(np.array([[0,1,0], [0,0,0], [0,0,0]]))
basic_input.append(np.array([[0,0,1], [0,0,0], [0,0,0]]))
basic_input.append(np.array([[0,0,0], [1,0,0], [0,0,0]]))
basic_input.append(np.array([[0,0,0], [0,1,0], [0,0,0]]))
basic_input.append(np.array([[0,0,0], [0,0,1], [0,0,0]]))
basic_input.append(np.array([[0,0,0], [0,0,0], [1,0,0]]))
basic_input.append(np.array([[0,0,0], [0,0,0], [0,1,0]]))
basic_input.append(np.array([[0,0,0], [0,0,0], [0,0,1]]))



basic_gt = []
basic_gt.append(0)
basic_gt.append(1)
basic_gt.append(2)
basic_gt.append(3)
basic_gt.append(4)
basic_gt.append(5)
basic_gt.append(6)
basic_gt.append(7)
basic_gt.append(8)



class Network(torch.nn.Module):
    def __init__(self):
        super(Network, self).__init__() 
        
        self.blocks = torch.nn.ModuleList([# sequential network blocks
                slayer.block.cuba.Input(neuronparams_cuba_inp),
                slayer.block.cuba.Conv(neuronparams_cuba, 1, 9,3,weight_norm=True, delay=False),
            ])

    def forward(self, x):
        
        for block in self.blocks:
            x = block(x)
        
        x = torch.squeeze(x)
        
        x = torch.mean(x, dim = 2)
        
        return x

    def export_hdf5(self, filename):
        # network export to hdf5 format
        h = h5py.File(filename, 'w')
        simulation = h.create_group('simulation')
        simulation['Ts'] = 1
        simulation['tSample'] = 20      
        layer = h.create_group('layer')
        for i, b in enumerate(self.blocks):
            b.export_hdf5(layer.create_group(f'{i}'))



                
 

thenet = Network()

optimizer = torch.optim.Adam(thenet.parameters(), lr = 0.1)

assistant = slayer.utils.Assistant(thenet.cuda(), torch.nn.CrossEntropyLoss(), optimizer, stats = slayer.utils.stats.LearningStats(), classifier=customClassifier)


epochs = 200

batchSize = 8

for epoch in range(epochs):

    batch_inp_list = []
    batch_gt_list = []
    for image in range(9):
        
        input = torch.unsqueeze(torch.tensor(basic_input[image]), dim = 0).float() 
        ground_truth = torch.ones(1).long()
        ground_truth[0] = basic_gt[image]

        batch_inp_list.append(input)
        batch_gt_list.append(ground_truth)
        
        if image % batchSize == 0 and image > 0 :

            input_batch = torch.stack(batch_inp_list, dim = 0)
            gt_batch = torch.squeeze(torch.stack(batch_gt_list, dim = 0))

            output = assistant.train(slayer.utils.time.replicate(input_batch, 40).cuda(), gt_batch.cuda())

            batch_inp_list = []
            batch_gt_list = []
            
    assistant.stats.print(epoch)
    assistant.stats.new_line()

    assistant.stats.update()     

thenet.export_hdf5('slayer_network.net')

```
Then use the following code to load the model and run it as a lava process 

```python
import os
import numpy as np
import typing as ty

# Import Process level primitives
from lava.magma.core.process.process import AbstractProcess
from lava.magma.core.process.variable import Var
from lava.magma.core.process.ports.ports import InPort, OutPort



basic_input = []
basic_input.append(np.array([[1,0,0], [0,0,0], [0,0,0]]))
basic_input.append(np.array([[0,1,0], [0,0,0], [0,0,0]]))
basic_input.append(np.array([[0,0,1], [0,0,0], [0,0,0]]))
basic_input.append(np.array([[0,0,0], [1,0,0], [0,0,0]]))
basic_input.append(np.array([[0,0,0], [0,1,0], [0,0,0]]))
basic_input.append(np.array([[0,0,0], [0,0,1], [0,0,0]]))
basic_input.append(np.array([[0,0,0], [0,0,0], [1,0,0]]))
basic_input.append(np.array([[0,0,0], [0,0,0], [0,1,0]]))
basic_input.append(np.array([[0,0,0], [0,0,0], [0,0,1]]))



basic_gt = []
basic_gt.append(0)
basic_gt.append(1)
basic_gt.append(2)
basic_gt.append(3)
basic_gt.append(4)
basic_gt.append(5)
basic_gt.append(6)
basic_gt.append(7)
basic_gt.append(8)

shape = (3,3,1)


np.set_printoptions(linewidth=np.inf)

class SpikeInput(AbstractProcess):
    """Reads image data from the MNIST dataset and converts it to spikes.
    The resulting spike rate is proportional to the pixel value."""

    def __init__(self,
                 vth: int,
                 num_images: ty.Optional[int] = 25,
                 num_steps_per_image: ty.Optional[int] = 128):
        super().__init__()
        
        self.spikes_out = OutPort(shape=shape)  # Input spikes to the classifier
        self.label_out = OutPort(shape=(1,))  # Ground truth labels to OutputProc
        self.num_images = Var(shape=(1,), init=num_images)
        self.num_steps_per_image = Var(shape=(1,), init=num_steps_per_image)
        self.input_img = Var(shape=shape)
        self.ground_truth_label = Var(shape=(1,))
        self.v = Var(shape=shape, init=0)
        self.vth = Var(shape=(1,), init=vth)


from lava.lib.dl.netx import hdf5

net = hdf5.Network(net_config='slayer_network.net')





class OutputProcess(AbstractProcess):
    """Process to gather spikes from 10 output LIF neurons and interpret the
    highest spiking rate as the classifier output"""

    def __init__(self, **kwargs):
        super().__init__()
        shape = (1,1,9)
        n_img = kwargs.pop('num_images', 25)
        self.num_images = Var(shape=(1,), init=n_img)
        self.spikes_in = InPort(shape=shape)
        self.label_in = InPort(shape=(1,))
        self.spikes_accum = Var(shape=shape)  # Accumulated spikes for classification
        self.num_steps_per_image = Var(shape=(1,), init=128)
        self.pred_labels = Var(shape=(n_img,))
        self.gt_labels = Var(shape=(n_img,))


# Import parent classes for ProcessModels
from lava.magma.core.model.sub.model import AbstractSubProcessModel
from lava.magma.core.model.py.model import PyLoihiProcessModel

# Import ProcessModel ports, data-types
from lava.magma.core.model.py.ports import PyInPort, PyOutPort
from lava.magma.core.model.py.type import LavaPyType

# Import execution protocol and hardware resources
from lava.magma.core.sync.protocols.loihi_protocol import LoihiProtocol
from lava.magma.core.resources import CPU

# Import decorators
from lava.magma.core.decorator import implements, requires

from lava.utils.dataloader.mnist import MnistDataset
#print(np.reshape(MnistDataset().images[1], (28,28)))
np.set_printoptions(linewidth=np.inf)


@implements(proc=SpikeInput, protocol=LoihiProtocol)
@requires(CPU)
class PySpikeInputModel(PyLoihiProcessModel):
    num_images: int = LavaPyType(int, int, precision=32)
    spikes_out: PyOutPort = LavaPyType(PyOutPort.VEC_DENSE, bool, precision=1)
    label_out: PyOutPort = LavaPyType(PyOutPort.VEC_DENSE, np.int32,
                                      precision=32)
    num_steps_per_image: int = LavaPyType(int, int, precision=32)
    input_img: np.ndarray = LavaPyType(np.ndarray, int, precision=32)
    ground_truth_label: int = LavaPyType(int, int, precision=32)
    v: np.ndarray = LavaPyType(np.ndarray, int, precision=32)
    vth: int = LavaPyType(int, int, precision=32)
    
    def __init__(self, proc_params):
        super().__init__(proc_params=proc_params)
        self.mnist_dataset = MnistDataset()
        self.curr_img_id = 0

    def post_guard(self):
        """Guard function for PostManagement phase.
        """
        if self.time_step % self.num_steps_per_image == 1:
            return True
        return False

    def run_post_mgmt(self):
        """Post-Management phase: executed only when guard function above 
        returns True.
        """
        img = np.expand_dims(basic_input[self.curr_img_id],axis = 2) 

        # In the current implementation the network will return the correct output if provided with the transpose of the input 
        # img = np.expand_dims(np.transpose(basic_input[self.curr_img_id]),axis = 2) 

        self.ground_truth_label = np.ones(1) * basic_gt[self.curr_img_id]
        print(self.ground_truth_label)
        self.input_img = img.astype(np.int32) #- 127
        self.v = np.zeros(self.v.shape)
        self.label_out.send(self.ground_truth_label)
        self.curr_img_id += 1

    def run_spk(self):
        """Spiking phase: executed unconditionally at every time-step
        """
        self.v = self.v + self.input_img
        s_out = self.v > self.vth
        self.v[s_out] = 0  # reset voltage to 0 after a spike
        self.spikes_out.send(s_out)

@implements(proc=OutputProcess, protocol=LoihiProtocol)
@requires(CPU)
class PyOutputProcessModel(PyLoihiProcessModel):
    label_in: PyInPort = LavaPyType(PyInPort.VEC_DENSE, int, precision=32)
    spikes_in: PyInPort = LavaPyType(PyInPort.VEC_DENSE, bool, precision=1)
    num_images: int = LavaPyType(int, int, precision=32)
    spikes_accum: np.ndarray = LavaPyType(np.ndarray, np.int32, precision=32)
    num_steps_per_image: int = LavaPyType(int, int, precision=32)
    pred_labels: np.ndarray = LavaPyType(np.ndarray, int, precision=32)
    gt_labels: np.ndarray = LavaPyType(np.ndarray, int, precision=32)
        
    def __init__(self, proc_params):
        super().__init__(proc_params=proc_params)
        self.current_img_id = 0

    def post_guard(self):
        """Guard function for PostManagement phase.
        """
        if self.time_step % self.num_steps_per_image == 0 and \
                self.time_step > 1:
            return True
        return False

    def run_post_mgmt(self):
        """Post-Management phase: executed only when guard function above 
        returns True.
        """
        gt_label = self.label_in.recv()
        print(self.spikes_accum)
        pred_label = np.argmax(self.spikes_accum)
        print(pred_label)
        self.gt_labels[self.current_img_id] = gt_label
        self.pred_labels[self.current_img_id] = pred_label
        self.current_img_id += 1
        self.spikes_accum = np.zeros_like(self.spikes_accum)

    def run_spk(self):
        """Spiking phase: executed unconditionally at every time-step
        """
        spk_in = self.spikes_in.recv()
        
        self.spikes_accum = self.spikes_accum + spk_in


num_images = 9
num_steps_per_image = 128

# Create Process instances
spike_input = SpikeInput(vth=1,
                         num_steps_per_image=num_steps_per_image)
mnist_clf = net
output_proc = OutputProcess(num_images=num_images)




# Connect Processes
spike_input.spikes_out.connect(mnist_clf.in_layer.neuron.a_in)
mnist_clf.out_layer.out.connect(output_proc.spikes_in)
# Connect Input directly to Output for ground truth labels
spike_input.label_out.connect(output_proc.label_in)



from lava.magma.core.run_conditions import RunSteps
from lava.magma.core.run_configs import Loihi1SimCfg
from lava.magma.core.sync.protocols.loihi_protocol import LoihiProtocol

# Loop over all images
for img_id in range(num_images):
    print(f"\rCurrent image: {img_id+1}", end="")

    # Run each image-inference for fixed number of steps
    mnist_clf.run(
        condition=RunSteps(num_steps=num_steps_per_image),
        run_cfg=Loihi1SimCfg(select_tag='fixed_pt'))




    mnist_clf.layers[0].neuron.u.set(np.zeros((2,2,1), dtype=np.int32))
    mnist_clf.layers[0].neuron.v.set(np.zeros((2,2,1), dtype=np.int32))

    mnist_clf.layers[1].neuron.u.set(np.zeros((1,1,9), dtype=np.int32))
    mnist_clf.layers[1].neuron.v.set(np.zeros((1,1,9), dtype=np.int32))



# Gather ground truth and predictions before stopping exec
ground_truth = output_proc.gt_labels.get().astype(np.int32)
predictions = output_proc.pred_labels.get().astype(np.int32)

print(output_proc.spikes_accum.get().astype(np.int32))




# Stop the execution
mnist_clf.stop()

accuracy = np.sum(ground_truth==predictions)/ground_truth.size * 100

print(f"\nGround truth: {ground_truth}\n"
      f"Predictions : {predictions}\n"
      f"Accuracy    : {accuracy}")

```
The lava process should produce the same output as the network it is based upon, but it does not.

**Environment:**
 - Device: Pc
 - OS: Linux
 - Lava version 0.8.0
 -  Lava-dl version 0.4.0

**Solution**
the issue can be fixed by changing line 486 in hdf5.py from 

```python
weight = weight.reshape(weight.shape[:4]).transpose((0, 3, 2, 1))
```
to

```python
weight = weight.reshape(weight.shape[:4]).transpose((0, 2, 3, 1))
```

**Comment by bamsumit** [comment]
@Karl-Vetter have you considered transposing your input instead?

**Comment by Karl-Vetter** [comment]
I have, as a result of past testing I thought that this does not work for multi layer networks. This is not correct, I in fact manged to produce networks for which neither method produced a network that functioned properly as a lava process. 
The methods also do not generally yield the same results. 
I think this issue can thus be closed as not being a bug, though the issue of poor performance after conversion remains. So far I have not found any instructions on how networks parameters ought to be set up for minimal loss of performance after conversion. Do any such instructions exist?

**Comment by bamsumit** [comment]
@Karl-Vetter  Please refer to the examples for
* https://github.com/lava-nc/lava-dl/blob/main/tutorials/lava/lib/dl/netx/pilotnet_sdnn/run.ipynb
* https://github.com/lava-nc/lava-dl/blob/main/tutorials/lava/lib/dl/netx/yolo_kp/run.ipynb
These are both convolutional network and you can find the examples for training to inference in this repo.

I am closing this issue for now.
